src/Chainweb/Pact/TransactionExec.hs

```diff
@@ -483,25 +488,71 @@ buyGas
     -> GasSupply
     -> ModuleCache
     -> IO (Either Text (T3 GasId [TxLog Value] ModuleCache))
-buyGas env cmd (Miner mid mks) supply mcache = do
-    let sender    = view (cmdPayload . pMeta . pmSender) cmd
-        initState = set (evalRefs . rsLoadedModules) mcache
-          $ initCapabilities [magic_FUND_TX]
+buyGas env cmd (Miner mid mks) supply mcache = go
+  where
+    sender = view (cmdPayload . pMeta . pmSender) cmd
+    initState = setModuleCache mcache $ initCapabilities [magic_FUND_TX]
+    interp = Interpreter $ \start end withRollback input ->
+      withRollback $ start (run input) >>= end
+    run input = do
+      put initState
+      findPayer >>= \r -> case r of
+        Nothing -> input
+        Just withPayerCap -> withPayerCap input
+
+    (Hash chash) = toUntypedHash (_cmdHash cmd)
+    bgHash = Hash (chash <> "-buygas")
+
+    go = do
+
+      buyGasCmd <- mkBuyGasCmd mid mks sender supply
+      -- print ("\n\nBUY GAS:",supply, "\n\n")
+      result <- applyExec' env interp buyGasCmd
+        (_pSigners $ _cmdPayload cmd) bgHash managedNamespacePolicy
+
+      let !mcache' = _erLoadedModules result
+
+      case _erExec result of
+        Nothing -> return $!
+          Left "buyGas: Internal error - empty continuation"
+        Just pe -> return $! Right $ T3 (GasId $ _pePactId pe) (_erLogs result) mcache'
+
+findPayer :: Eval e (Maybe (RunEval e -> RunEval e))
+findPayer = return Nothing -- disabled until tested
+
+_findPayer :: Eval e (Maybe (RunEval e -> RunEval e))
+_findPayer = runMaybeT go
+  where
+    go = do
+      (m,qn,as) <- MaybeT findPayerCap
+      pMod <- MaybeT $ lookupModule qn m
+      MaybeT $ return $ validateMod pMod
+      capRef <- MaybeT $ resolveRef qn (QName qn)
+      return $ runCap (getInfo qn) capRef as
+
+    findPayerCap :: Eval e (Maybe (ModuleName,QualifiedName,[PactValue]))
+    findPayerCap = preview $ eeMsgSigs . folded . folded . to sigPayerCap . _Just
+
+    sigPayerCap (UserCapability q@(QualifiedName m n _) as) | n == "GAS_PAYER" = Just (m,q,as)
```

Comment:
Shouldn't we also check the module name here to avoid collision? This needs to be `gas-payer-v1-blah.GAS_PAYER` doesn't it?

Reply:
We validate the module next. It doesn't really matter what it is as long as it implements the interface, which we check.